### PR TITLE
LLL: Tests for the alloc built-in.

### PR DESCRIPTION
**Pull request by benjaminion**
I think I now understand the alloc built-in :-)

### CODE REVIEW

File: test/liblll/EndToEndTest.cpp [hunk 1]
```diff
@@ -386,6 +386,30 @@ BOOST_AUTO_TEST_CASE(keccak256_32bytes)
...
+	BOOST_CHECK(callFallback() == encodeArgs(u256(64)));
+}
+
+BOOST_AUTO_TEST_CASE(alloc_2)
+{
+	char const* sourceCode = R"(
+		(returnlll
+			(seq
+				(mstore 0x40 0xff)
+				(mstore8 (alloc 1) 42)
```
Comment (axic) [inline test/liblll/EndToEndTest.cpp]: I think this here should assign the value of `(alloc 1)` to a variable and use that in return as the starting address.

it would be nice to have another test which ensures the starting address equals to what it should be. For that maybe the test you have is enough.

Comment (benjaminion) [reply test/liblll/EndToEndTest.cpp]: Sure.

Actually, I think I have found an edge-case bug in alloc.  If you do (alloc 0) it will corrupt the previous word in memory.  I'll raise an issue about this later.

**Comment by benjaminion** [comment]
I've updated the tests to be a bit more nuanced, and to combine with variable setting as per the suggestion.

Some tests are commented pending a plan for Issue #2465.

File: test/liblll/EndToEndTest.cpp [hunk 2]
```diff
@@ -386,6 +386,60 @@ BOOST_AUTO_TEST_CASE(keccak256_32bytes)
...
+	BOOST_CHECK(callContractFunction("test()", 32) == encodeArgs(u256(32)));
+	BOOST_CHECK(callContractFunction("test()", 33) == encodeArgs(u256(64)));
+}
+
+BOOST_AUTO_TEST_CASE(alloc_start)
+{
+	char const* sourceCode = R"(
+		(returnlll
+			(seq
+				(mstore 0x40 0)
```
Comment (axic) [inline test/liblll/EndToEndTest.cpp:473]: Why is this needed?

Comment (benjaminion) [reply test/liblll/EndToEndTest.cpp:473]: This is only to set msize initially to something other than 0x00 so we're testing a slightly more interesting case. After this `(msize)` and hence the `(alloc 0x01)` should both evaluate to 0x60.  If we don't do this it comes out to 0x00, which is duller.

More importantly, the point is to catch any case where `alloc` gets modified in such a way that it starts to ignore msize (for example).

**Comment by axic** [comment]
@benjaminion can you please rebase this?

**Comment by benjaminion** [comment]
Rebased.

File: test/liblll/EndToEndTest.cpp [hunk 3]
```diff
@@ -436,6 +436,60 @@ BOOST_AUTO_TEST_CASE(send_three_args)
...
+				(return (alloc 0x01))))
+	)";
+	compileAndRun(sourceCode);
+	BOOST_CHECK(callFallback() == encodeArgs(fromHex("60")));
```
Comment (axic) [inline test/liblll/EndToEndTest.cpp:477]: Shouldn't `encodeArgs(u256(0x60))` work here?

**Comment by axic** [comment]
I'd say let's fix `alloc 0` as discussed (once decided) in #2465 here.

File: test/liblll/EndToEndTest.cpp (hunk 3 above)
Comment (benjaminion) [reply test/liblll/EndToEndTest.cpp:477]: Yes, that should also work :-)

I'll update this when we've settled on the final approach.

**Comment by axic** [comment]
Replaced by #2545.
